server/src/main/java/org/elasticsearch/search/fetch/subphase/InnerHitsPhase.java

```diff
@@ -84,7 +88,24 @@ private void hitExecute(Map<String, InnerHitsContext.InnerHitSubContext> innerHi
             if (results == null) {
                 hit.setInnerHits(results = new HashMap<>());
             }
-            innerHitsContext.queryResult().topDocs(topDoc, innerHitsContext.sort() == null ? null : innerHitsContext.sort().formats);
+
+            final DocValueFormat[] sortValueFormats = innerHitsContext.sort() == null ? null : innerHitsContext.sort().formats;
+            List<RescoreContext> rescore = innerHitsContext.rescore();
+            if (rescore.isEmpty() == false) {
+                assert sortValueFormats == null;
+
+                TopDocs justTopDocs = topDoc.topDocs;
+                if (justTopDocs.scoreDocs.length > 0) {
+                    for (RescoreContext ctx : rescore) {
+                        justTopDocs = ctx.rescorer().rescore(justTopDocs, innerHitsContext.searcher(), ctx);
+                        // TODO: Assert that top docs are sorted by score
+                    }
+
+                    topDoc = new TopDocsAndMaxScore(justTopDocs, justTopDocs.scoreDocs[0].score);
+                }
+            }
+
+            innerHitsContext.queryResult().topDocs(topDoc, sortValueFormats);
```

Comment:
I think this will throw an NPE if you pass in a null `sortValueFormats`

Reply:
`sortValueFormats` is `null` when sorting by score. It may appear at first glance that `topDocs` will throw an NPE when `sortValueFormats` is `null`, but `sortValueFormats.length` is only called when the sort is configured to sort on a field (hence the `instanceof FieldDoc` check). In that case, `sortValueFormats` should not be `null`.